### DISCUSSION TITLE
How do I get a BLEManager instance?

### DISCUSSION THREAD

**Discussion by davepruitt** [shinyorg/shiny#1391] [Q&A]
The very first thing in the documentation (https://shinylib.net/client/ble/manager/) says that you need to resolve/inject/etc an instance of IBLEManager to do _anything_ with BLE. But there is **no** documentation on how to do this. How?!?! I've dived deep into the sample code, but I have found nothing.

**Comment by aritchie** [comment]
It depends on how you setup your dependency injection.  At the viewmodel level, at the page level... it is hard to say how far people go here.  You can use:

```csharp
var manager = Shiny.Hosting.Host.GetService<IBleManager>();
```

Sample with DI is here: https://github.com/shinyorg/shiny/blob/dev/samples/Sample.Maui/BleClient/ScanViewModel.cs

**Reply by davepruitt** [reply]
Your sample with dependency injection does not actually show _how_ to get the BLEManager object anywhere.

Your one-line code sample in your post above does:
```
var manager = Shiny.Hosting.Host.GetService<IBleManager>();
```

I suggest this line of code should be added to the documentation, because I banged my head against my desk for hours trying to figure out how to get this object.

**Reply by aritchie** [reply]
This isn't a product.  I do the best I can with the time I have.  My samples have full MAUI DI working, but there are just too many scenarios out there and MAUI is my main use-case for now.  

Feel free to send a PR with doc updates.

**Reply by aritchie** [reply]
The sample does show how to get the IBleManager in the link I copied.  It is in the constructor because of the MAUI DI.

**Reply by davepruitt** [reply]
The dependency injection doesn't show the user of this repository how to actually get the object. It just passes that responsibility on to the class that instantiates this class. At some point the object has to be created somewhere in the code. I understand dependency injection and I use it, but there still needs to be somewhere in the code that actually creates the object.

**Reply by aritchie** [reply]
I'm not looking to argue about your opinions here.  I gave you an answer to solve your problem and instead of a "thank you" or just moving along, you send in a complaint.  If you don't like the docs or the library or it causes you to "bang your head for hours", just use something else.

**Reply by davepruitt** [reply]
Sorry if my response sounded like a complaint. While I do appreciate your efforts, I was simply pointing out that the documentation needs that single line of code to really get people using the BLE library:
```
var manager = Shiny.Hosting.Host.GetService<IBleManager>();
```

Without that line of code, there is no way to actually create/get an instance of the BLEManager object which I can then pass around to other places within the application. Therefore that is "point 0" where everything needs to begin. Not having that line of code in the documentation is very limiting.

**Reply by aritchie** [reply]
As I said, it is built into MAUI.  You can inject your viewmodels/views/etc with IBleManager if they are part of the container.  That's just how dependency injection works especially with in MAUI.  I don't really want to explain the entire mechanism to every user that comes through.  

If I put your specific example, people will start complaining about purest DI, unit testing, etc, so I just don't.  I can't win and won't even try to win with everyone, but I have heard your feedback.

Best of luck with your project.
